Add AmountInput interface specification (#15)
tonlabs/DeBot-IS-consortium: AmountInput/README.md

# AmountInput Interface

**Status**: Proposed

| Name        | ID                                                                |
| :---------- | :---------------------------------------------------------------- |
| AmountInput | a1d347099e29c1624c8890619daf207bde18e92df5220a54bcc6d858309ece84  |

## Description

Interface for amount input.

## Functions

> `get` - prints prompt message to the user and returns number of nanotons entered by user.

Reviewer: ```suggestion
`get` - prints prompt message to the user and returns amount of tokens entered by user.
```

arguments:

	answerId: uint32
    prompt: bytes

>     step: uint256

Reviewer 2: What is step? I believe that all arguments should be documented properly.

> 	min: uint256

Reviewer 2: What about negative values? Aren't they possible?

Reviewer: Do we need a negative amount of tokens?

Reviewer 2: Don't know, but I can imagine an example of such a case. Maybe it's redundant indeed.

Author: I think that negative number is incorrect param.

> 	max: uint256

Reviewer 2: Same concern

> 	answerId: uint32
>     prompt: bytes
>     step: uint256
> 	min: uint256
> 	max: uint256

Reviewer: ```suggestion
    answerId: uint32 - id of callback function.
    prompt: bytes - utf-8 string to print to the user before input.
    decimals: uint8 - number decimals that input currency has, e.g. 9 - means browser should multiply the amount to 1000000000 to get its debot representation.
    min: uint256 - minimal allowed amount.
    max: uint256 - maximum allowed amount.
```

Reviewer: [amount_input.tar.gz](https://github.com/tonlabs/DeBot-IS-consortium/files/5981375/amount_input.tar.gz)

AmountInput.sol
AmountInput.abi.json
AmountInput.hpp
Example.sol

returns:

> 	{ value: uint256 }

Reviewer 2: Same concern

## Declaration in Solidity

> TODO: Add declaration

Reviewer: ```suggestion
interface IAmountInput {

	function get(uint32 answerId, string prompt, uint8 decimals, uint256 min, uint256 max) external returns (uint256 value);

}
```

## Declaration in C++

> TODO: Add declaration

Reviewer: ```suggestion
namespace tvm { namespace schema {

__interface IAmountInput {
	[[internal, answer_id]]
	uint256 get(string prompt, uint8 decimals, uint256 min, uint256 max);
}

}};
```

## Code Example

### Solidity

> TODO: Add example

Reviewer: ```suggestion
Example.sol
```

# AmountInput Interface

**Status**: Proposed

| Name        | ID                                                                |
| :---------- | :---------------------------------------------------------------- |
| AmountInput | a1d347099e29c1624c8890619daf207bde18e92df5220a54bcc6d858309ece84  |

## Description

Interface for amount input.

## Functions

`get` - prints prompt message to the user and returns amount of tokens entered by user.

arguments:

> 	answerId: uint32
> 	prompt: bytes
> 	decimal: uint8
> 	min: uint256
> 	max: uint256

Reviewer: ```suggestion
	answerId: uint32 - id of function callback.
	prompt: bytes - utf-8 string to print to the user before input.
	decimal: uint8 - number of decimals used by amount token, e.g. 9 - means that browser should multiply token amount to 1000000000 to get its debot represenation.
	min: uint256 - minimal amount that can be entered by user.
	max: uint256 - maximum amount that can be entered by user.
```

> 	min: uint256 - minimal amount that can be entered by user.

Reviewer 2: Need to clarify to the developer that this is a value multiplied by 10^decimal

Reviewer: Done

returns:

> 	{ value: uint256 }

Reviewer 2: Need to clarify to developer that the value is multiplied by 10^decimal here as well

Reviewer: Done

> 	decimal: uint8 - number of decimals used by amount token, e.g. 9 - means that browser should multiply token amount to 1000000000 to get its debot represenation.
> 	min: uint256 - minimal amount that can be entered by user.
> 	max: uint256 - maximum amount that can be entered by user.
> 
> returns:
> 
> 	{ value: uint256 }

Author: ```suggestion
	decimal: uint8 - number of decimals used by amount token, e.g. 9 - means that browser should multiply token amount to 1000000000 to get its debot represenation.
	min: uint128 - minimal amount that can be entered by user.
	max: uint128 - maximum amount that can be entered by user.

returns:

	{ value: uint128 }
```
@Keshoid Approve?

Reviewer: Yep

> 	{ value: uint256 }

Reviewer: ```suggestion
	{ value: uint256 } - amount of tokens multiplied by `10^decimals`. To get the integer part of an amount debot should divide value by `10^decimals`.
```

> 	{ value: uint128 }

Author: ```suggestion
	{ value: uint128 } - amount of tokens multiplied by `10^decimals`. To get the integer part of an amount debot should divide value by `10^decimals`.
```

## Declaration in Solidity

interface IAmountInput {

> 	function get(uint32 answerId, string prompt, uint8 decimals, uint256 min, uint256 max) external returns (uint256 value);
> 
> }
> 
> ## Declaration in C++
> 
> namespace tvm { namespace schema {
> 
> __interface IAmountInput {
> 	[[internal, answer_id]]
> 	uint256 get(string prompt, uint8 decimals, uint256 min, uint256 max);

Author: ```suggestion
	function get(uint32 answerId, string prompt, uint8 decimals, uint128 min, uint128 max) external returns (uint128 value);

}

## Declaration in C++

namespace tvm { namespace schema {

__interface IAmountInput {
	[[internal, answer_id]]
	uint128 get(string prompt, uint8 decimals, uint128 min, uint128 max);
```
@Keshoid Approve?

Reviewer: Yep
